api/ai/new/double_round_robin_algorithm/double_round_robin_algorithm.py

```diff
@@ -0,0 +1,128 @@
+from api.ai.new.double_round_robin_algorithm.utils import calculate_utilities
+from api.models.student import Student
+from api.models.team import Team
+from api.models.team_set import TeamSet
+
+
+class DoubleRoundRobinAlgorithm(Algorithm):
+    def __init__(
+        self,
+        algorithm_options: DoubleRoundRobinAlgorithmOptions,
+        team_generation_options: TeamGenerationOptions,
+    ):
+        super().__init__(algorithm_options, team_generation_options)
+
+        self.projects = algorithm_options.projects
+
+    def _round_robin(
+        self,
+        utilities: Dict[int, List[Utility]],
+        num_students: int,
+        allocation: Dict[int, List[Student]],
+        selected_students: Set[int],
+    ) -> Dict[int, List[Student]]:
+        while len(selected_students) < num_students:
+            for project in self.projects:
+                while (
+                    len(utilities[project.id]) > 0
+                    and utilities[project.id][0].student.id in selected_students
+                ):
+                    utilities[project.id].pop(0)
+
+                if len(utilities[project.id]) == 0:
+                    continue
+
+                curr_utility = utilities[project.id].pop(0)
+                allocation[project.id].append(curr_utility.student)
+                selected_students.add(curr_utility.student.id)
+
+        return allocation
+
+    def generate(self, students: List[Student]) -> TeamSet:
+        utilities = calculate_utilities(self.projects, students)
+
+        positive_utilities: Dict[int, List[Utility]] = {
+            project.id: [] for project in self.projects
+        }
+        positive_utility_students: Set[int] = set()
+        negative_utilities: Dict[int, List[Utility]] = {
+            project.id: [] for project in self.projects
+        }
+        negative_utility_students: Set[int] = set()
+
+        for project in self.projects:
+            for student in students:
+                curr_utility = utilities[project.id][student.id]
+                if curr_utility.value > 0:
+                    heappush(positive_utilities[project.id], curr_utility)
+                    positive_utility_students.add(student.id)
+                else:
+                    heappush(negative_utilities[project.id], curr_utility)
+                    negative_utility_students.add(student.id)
+
+        selected_students: Set[int] = set()
+        allocation: Dict[int, List[Student]] = {
+            project.id: [] for project in self.projects
+        }
+        self._round_robin(
+            negative_utilities,
+            len(negative_utility_students),
+            allocation,
+            selected_students,
+        )
+        self._round_robin(
+            positive_utilities,
+            len(positive_utility_students),
+            allocation,
+            selected_students,
+        )
+
+        team_list: List[Team] = []
```

Comment:
try running this and right before this just `print(self.teams)` to see what happens, I'm curious about something